### PR TITLE
Add command to export purged records to CSV files

### PR DESCRIPTION
**Pull request by ephraimbuddy**
This adds `airflow db export-cleaned --export-format csv --output-path /some/path` command
to export records from the archived tables to csv files

### CODE REVIEW

File: airflow/utils/db_cleanup.py [hunk 1]
```diff
@@ -159,6 +170,14 @@ def _do_delete(*, query, orm_model, skip_archive, session):
     logger.debug("delete statement:\n%s", delete.compile())
     session.execute(delete)
     session.commit()
+    if export_to_csv:
+        if not output_path.startswith(AIRFLOW_HOME):
+            output_path = os.path.join(AIRFLOW_HOME, output_path)
```
Comment (jedcunningham) [inline airflow/utils/db_cleanup.py]: We shouldn't restrict this to AIRFLOW_HOME.

File: airflow/utils/db_cleanup.py [hunk 2]
```diff
@@ -333,6 +362,8 @@ def run_cleanup(
     :param verbose: If true, may provide more detailed output.
     :param confirm: Require user input to confirm before processing deletions.
     :param skip_archive: Set to True if you don't want the purged rows preservied in an archive table.
+    :param export_to_csv: Set to True if you want the purged rows exported to a CSV file.
+    :param output_path: Path to the directory where the CSV file will be written.
```
Comment (jedcunningham) [inline airflow/utils/db_cleanup.py]:
```suggestion
    :param output_path: Path to the directory where the CSV files will be written.
```

File: airflow/cli/cli_parser.py [hunk 3]
```diff
@@ -468,7 +468,18 @@ def string_lower_type(val):
     help="Don't preserve purged records in an archive table.",
     action="store_true",
 )
-
+ARG_DB_EXPORT_TO_CSV = Arg(
```
Comment (jedcunningham) [inline airflow/cli/cli_parser.py]: I feel like we should merge this with `output-path`, which should be required.

File: airflow/utils/db_cleanup.py [hunk 4]
```diff
@@ -123,7 +126,15 @@ def _check_for_rows(*, query: Query, print_rows=False):
     return num_entities
 
 
-def _do_delete(*, query, orm_model, skip_archive, session):
+def _to_csv(*, target_table, file_path, session):
+    with open(file_path, "w") as f:
```
Comment (jedcunningham) [inline airflow/utils/db_cleanup.py]: This will silently overwrite any existing file, probably should skip/fail instead?

Comment (ephraimbuddy) [reply airflow/utils/db_cleanup.py]: I don't think there would be a case of overwriting since the files are based on timestamps.

File: airflow/utils/db_cleanup.py [hunk 5]
```diff
@@ -159,6 +170,14 @@ def _do_delete(*, query, orm_model, skip_archive, session):
...
     session.execute(delete)
     session.commit()
+    if export_to_csv:
+        if not output_path.startswith(AIRFLOW_HOME):
+            output_path = os.path.join(AIRFLOW_HOME, output_path)
+        os.makedirs(output_path, exist_ok=True)
+        _to_csv(
+            target_table=target_table, file_path=f"{output_path}/{target_table_name}.csv", session=session
+        )
+        skip_archive = True
```
Comment (dstandish) [inline airflow/utils/db_cleanup.py]: i reckon we should probably not force skip archive just because user wants csv output.  user can add skip archive also. wdyt?

File: airflow/utils/db_cleanup.py (hunk 1 above)
Comment (dstandish) [reply airflow/utils/db_cleanup.py]: yeah, probably we should just accept whatever path the user has requested.  i guess depending on whether we want to create the parent dirs for the user if they don't exist, we could either do so, or fail fast if they don't

File: airflow/cli/cli_parser.py (hunk 3 above)
Comment (dstandish) [reply airflow/cli/cli_parser.py]: or what if we have output path and format. we could support only csv at the moment but json down the road.
perhaps output path can optionally be `-` for stdout.

File: airflow/utils/db_cleanup.py (hunk 5 above)
Comment (ephraimbuddy) [reply airflow/utils/db_cleanup.py]: I think it depends on how you look at this. If `--export-to-csv` means exporting what is in the archive then explicit skip(`--skip-archive`) sounds good. However, if `--export-to-csv` means to archive data in csv, then forcing skip archive sounds good to me since the data is now archived in csv files. 
I had this discussion with @norm, also @jedcunningham mentioned this too when we talked about the idea. WDYT: @norm @jedcunningham

File: airflow/cli/cli_parser.py (hunk 3 above)
Comment (ephraimbuddy) [reply airflow/cli/cli_parser.py]: Maybe we shouldn't combine it but have e.g `--export-to` or `--archive-to` that defaults to `database` with other possible options like `csv`, `json` etc? `csv` at the moment of course.

Comment (jedcunningham) [reply airflow/cli/cli_parser.py]: That works for me. Json really appeals to me since you could jam it all in a single file, which also makes stdout much easier.

File: airflow/utils/db_cleanup.py (hunk 4 above)
Comment (jedcunningham) [reply airflow/utils/db_cleanup.py]: Ah, I totally missed the timestamp portion! Thanks!

File: airflow/utils/db_cleanup.py (hunk 5 above)
Comment (jedcunningham) [reply airflow/utils/db_cleanup.py]: I definitely lean toward having it go to one place.

File: airflow/cli/cli_parser.py (hunk 3 above)
Comment (ephraimbuddy) [reply airflow/cli/cli_parser.py]: I poked around `json` and there are issues on serialization. I think archiving to `json` should be done in a separate PR

Comment (dstandish) [reply airflow/cli/cli_parser.py]: Yeah i am not suggesting you implement json now... only suggesting that we have separate "out file" from "format" params

File: airflow/utils/db_cleanup.py (hunk 5 above)
Comment (dstandish) [reply airflow/utils/db_cleanup.py]: I just think it doesn't really hurt to leave the data backed up in a table.  A user might not understand that by wanting the data exported it will also be deleted irretrievably.  What if something goes wrong with the export?  What if the user is running this with `kubectl exec -ti` and the pod dies before they save the data?  I would let skip archive define the skip archive behavior _always_ and not have the export behavior determine it.   User can always drop the tables later but they can't bring them back.  We also can't really sure that our CSV export would actually be easily importable back into the database without loss.

BUT that's of course just one man's take

Comment (jedcunningham) [reply airflow/utils/db_cleanup.py]: Even with the data in those tables it isn't trivial to put it back, potentially (e.g. backfill conflicts). But yes, point taken. Maybe we should have a cli command to drop the tables. That keeps it "easy".

**Review by norm — CHANGES_REQUESTED** [review]
Think `docs/apache-airflow/howto/usage-cli.rst` needs to be updated too.

File: airflow/utils/db_cleanup.py (hunk 5 above)
Comment (ephraimbuddy) [reply airflow/utils/db_cleanup.py]: I thought about this more and I think we should change the approach....instead of running the  export right when we are cleaning data which to me means archiving to file, what do you think about having a separate command `airflow db export-cleaned` that would export from the archive tables to the given output folder? I experimented with this, see below:

```diff
diff --git a/airflow/cli/cli_parser.py b/airflow/cli/cli_parser.py
index 6890368f9b..4e260da7c5 100644
--- a/airflow/cli/cli_parser.py
+++ b/airflow/cli/cli_parser.py
@@ -468,6 +468,17 @@ ARG_DB_SKIP_ARCHIVE = Arg(
     help="Don't preserve purged records in an archive table.",
     action="store_true",
 )
+ARG_DB_EXPORT_FORMAT = Arg(
+    ("--export-format",),
+    help="The file format to export the cleaned data",
+    choices=("csv",)
+)
+ARG_DB_OUTPUT_PATH = Arg(
+    ("--output-path",),
+    help="The path to export the cleaned data",
+    metavar='FILEPATH',
+    required=True,
+)
 
 
 # pool
@@ -1589,6 +1600,15 @@ DB_COMMANDS = (
             ARG_DB_SKIP_ARCHIVE,
         ),
     ),
+    ActionCommand(
+        name="export-cleaned",
+        help="Export purged records from the archive tables",
+        func=lazy_load_command("airflow.cli.commands.db_command.export_archived"),
+        args=(
+            ARG_DB_EXPORT_FORMAT,
+            ARG_DB_OUTPUT_PATH,
+        )
+    )
 )
 CONNECTIONS_COMMANDS = (
     ActionCommand(
diff --git a/airflow/cli/commands/db_command.py b/airflow/cli/commands/db_command.py
index 4064285db8..d27a569e35 100644
--- a/airflow/cli/commands/db_command.py
+++ b/airflow/cli/commands/db_command.py
@@ -27,7 +27,7 @@ from airflow import settings
 from airflow.exceptions import AirflowException
 from airflow.utils import cli as cli_utils, db
 from airflow.utils.db import REVISION_HEADS_MAP
-from airflow.utils.db_cleanup import config_dict, run_cleanup
+from airflow.utils.db_cleanup import config_dict, run_cleanup, export_cleaned
 from airflow.utils.process_utils import execute_interactive
 
 
@@ -207,3 +207,13 @@ def cleanup_tables(args):
         confirm=not args.yes,
         skip_archive=args.skip_archive,
     )
+
+
+@cli_utils.action_cli(check_db=False)
+def export_archived(args):
+    if not os.path.exists(args.output_path):
+        raise AirflowException(f"The specified --output-path {args.output_path} does not exist")
+    export_cleaned(
+        export_format=args.export_format,
+        output_path=args.output_path,
+    )
diff --git a/airflow/utils/db_cleanup.py b/airflow/utils/db_cleanup.py
index ac96a6abdb..cbb5963db1 100644
--- a/airflow/utils/db_cleanup.py
+++ b/airflow/utils/db_cleanup.py
@@ -20,13 +20,14 @@ This module took inspiration from the community maintenance dag
 """
 from __future__ import annotations
 
+import csv
 import logging
 from contextlib import contextmanager
 from dataclasses import dataclass
 from typing import Any
 
 from pendulum import DateTime
-from sqlalchemy import and_, column, false, func, table, text
+from sqlalchemy import and_, column, false, func, table, text, inspect
 from sqlalchemy.exc import OperationalError, ProgrammingError
 from sqlalchemy.ext.compiler import compiles
 from sqlalchemy.orm import Query, Session, aliased
@@ -40,6 +41,8 @@ from airflow.utils.session import NEW_SESSION, provide_session
 
 logger = logging.getLogger(__file__)
 
+ARCHIVE_TABLE_PREFIX = "_airflow_deleted__"
+
 
 @dataclass
 class _TableConfig:
@@ -131,7 +134,7 @@ def _do_delete(*, query, orm_model, skip_archive, session):
     # using bulk delete
     # create a new table and copy the rows there
     timestamp_str = re.sub(r"[^\d]", "", datetime.utcnow().isoformat())[:14]
-    target_table_name = f"_airflow_deleted__{orm_model.name}__{timestamp_str}"
+    target_table_name = f"{ARCHIVED_TABLE_PREFIX}{orm_model.name}__{timestamp_str}"
     print(f"Moving data to table {target_table_name}")
     stmt = CreateTableAs(target_table_name, query.selectable)
     logger.debug("ctas query:\n%s", stmt.compile())
@@ -370,3 +373,27 @@ def run_cleanup(
                 session=session,
             )
             session.commit()
+
+
+def _dump_db(*, target_table, file_path, export_format, session):
+    if export_format == "csv":
+        with open(file_path + ".csv", "w") as f:
+            csv_writer = csv.writer(f)
+            cursor = session.execute(text(f"SELECT * FROM {target_table}"))
+            csv_writer.writerow(cursor.keys())
+            csv_writer.writerows(cursor.fetchall())
+
+
+@provide_session
+def export_cleaned(export_format, output_path, session: Session = NEW_SESSION):
+    inspector = inspect(session.bind)
+    table_names = inspector.get_table_names()
+    for table_name in table_names:
+        if table_name.startswith(ARCHIVE_TABLE_PREFIX):
+            _dump_db(
+                target_table=table_name,
+                file_path=f"{output_path}/{table_name}",
+                export_format=export_format,
+                session=session,
+            )
+    logger.info("Done")
```

Comment (dstandish) [reply airflow/utils/db_cleanup.py]: I like it @ephraimbuddy

Comment (dstandish) [reply airflow/utils/db_cleanup.py]: And then we could maybe at some point add an interactive "drop archive tables" option or subcommand which would allow user to select which tables to drop.  Something like that

Comment (norm) [reply airflow/utils/db_cleanup.py]: Gives users a nice clean, export, then drop action that reduces the possibility of something going wrong and losing data should you want to keep backups.

Comment (ephraimbuddy) [reply airflow/utils/db_cleanup.py]: Fixed

**Review by norm — APPROVED** [review]

File: airflow/cli/cli_parser.py [hunk 6]
```diff
@@ -468,7 +468,17 @@ def string_lower_type(val):
...
+ARG_DB_EXPORT_FORMAT = Arg(
+    ("--export-format",),
+    help="The file format to export the cleaned data",
+    choices=("csv",),
+)
+ARG_DB_OUTPUT_PATH = Arg(
+    ("--output-path",),
+    metavar="FILEPATH",
+    help="The output to export the cleaned data",
+    required=True,
```
Comment (ephraimbuddy) [inline airflow/cli/cli_parser.py:490]: Should we have a default directory instead of making this a required option?

Comment (dstandish) [reply airflow/cli/cli_parser.py:490]: I think required makes sense

File: airflow/utils/db_cleanup.py [hunk 7]
```diff
@@ -370,3 +382,19 @@ def run_cleanup(
                 session=session,
             )
             session.commit()
+
+
+@provide_session
+def export_archived_records(export_format, output_path, session: Session = NEW_SESSION):
```
Comment (kaxil) [inline airflow/utils/db_cleanup.py]: Hmm .. Should we instead have a flag on `db clean` command to `export` before clean/archive? Or do we expect users to run this command separately to db clean too?

Comment (ephraimbuddy) [reply airflow/utils/db_cleanup.py]: That was the first thing we did but it looks like a separate command is better...pushing all the contents of the archived table to filesystem. See https://github.com/apache/airflow/pull/29058#discussion_r1087901858

Comment (kaxil) [reply airflow/utils/db_cleanup.py]: Cool, I will leave that upto you

File: airflow/cli/commands/db_command.py [hunk 8]
```diff
@@ -207,3 +207,14 @@ def cleanup_tables(args):
         confirm=not args.yes,
         skip_archive=args.skip_archive,
     )
+
+
+@cli_utils.action_cli(check_db=False)
+def export_archived(args):
+    """Exports archived records from metadata database."""
+    if not os.path.exists(args.output_path):
+        raise AirflowException(f"The specified --output-path {args.output_path} does not exist.")
```
Comment (uranusjr) [inline airflow/cli/commands/db_command.py]: Probably should be an `isdir` check instead? (Also may be a good idea to mention this should be a directory in the help message and documentation; it took me a minute to figure that out.)

File: airflow/utils/db_cleanup.py [hunk 9]
```diff
@@ -370,3 +382,19 @@ def run_cleanup(
...
+    """Export cleaned data to the given output path in the given format."""
+    inspector = inspect(session.bind)
+    table_names = inspector.get_table_names()
+    for table_name in table_names:
+        if table_name.startswith(ARCHIVE_TABLE_PREFIX):
+            _dump_db(
+                target_table=table_name,
+                file_path=f"{output_path}/{table_name}",
```
Comment (uranusjr) [inline airflow/utils/db_cleanup.py]: Better to use `os.path.join`. Why is the `.csv` suffix addition not done here?

File: airflow/utils/db_cleanup.py [hunk 10]
```diff
@@ -123,6 +126,15 @@ def _check_for_rows(*, query: Query, print_rows=False):
     return num_entities
 
 
+def _dump_db(*, target_table, file_path, export_format, session):
```
Comment (uranusjr) [inline airflow/utils/db_cleanup.py]: Should raise an exception if `export_format` is an unexpected value (future-proof).

Comment (ephraimbuddy) [reply airflow/utils/db_cleanup.py]: Added a test in argparse since the choices=('csv',) would prevent the incorrect format from making its way to this function

File: airflow/cli/commands/db_command.py (hunk 8 above)
Comment (ephraimbuddy) [reply airflow/cli/commands/db_command.py]: moved the check to parser

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]
Nice little command :)

File: airflow/cli/cli_parser.py [hunk 11]
```diff
@@ -1589,6 +1608,15 @@ class GroupCommand(NamedTuple):
             ARG_DB_SKIP_ARCHIVE,
         ),
     ),
+    ActionCommand(
+        name="export-cleaned",
+        help="Export cleaned data from the archive tables",
+        func=lazy_load_command("airflow.cli.commands.db_command.export_archived"),
```
Comment (jedcunningham) [inline airflow/cli/cli_parser.py]: We have `export-cleaned` and `export_archived`, we should pick one.

File: airflow/utils/db_cleanup.py [hunk 12]
```diff
@@ -123,6 +127,15 @@ def _check_for_rows(*, query: Query, print_rows=False):
...
```
Comment (jedcunningham) [inline airflow/utils/db_cleanup.py]:
```suggestion
def _dump_table_to_file(*, target_table, file_path, export_format, session):
```

nit

File: airflow/utils/db_cleanup.py [hunk 13]
```diff
@@ -370,3 +383,19 @@ def run_cleanup(
...
+    inspector = inspect(session.bind)
+    table_names = inspector.get_table_names()
+    for table_name in table_names:
+        if table_name.startswith(ARCHIVE_TABLE_PREFIX):
+            _dump_db(
+                target_table=table_name,
+                file_path=os.path.join(output_path, f"{table_name}.{export_format}"),
+                export_format=export_format,
+                session=session,
+            )
```
Comment (jedcunningham) [inline airflow/utils/db_cleanup.py]: I feel like we should have a flag to drop the table as well?

Comment (ephraimbuddy) [reply airflow/utils/db_cleanup.py]: Also thinking of adding table names so one can export specific archive tables. WDYT?

Comment (jedcunningham) [reply airflow/utils/db_cleanup.py]: SGMT, sorta like we do on the clean command.

File: airflow/utils/db_cleanup.py [hunk 14]
```diff
@@ -123,6 +127,15 @@ def _check_for_rows(*, query: Query, print_rows=False):
     return num_entities
 
 
+def _dump_table_to_file(*, target_table, file_path, export_format, session):
+    if export_format == "csv":
```
Comment (norm) [inline airflow/utils/db_cleanup.py:132]: It's a minor thing since you *shouldn't* be able to get here without `csv` being the format, but there's no code to catch what-if-no-acceptable-format.

(I mean, "you shouldn't see this" is a classic comedy error message for a reason, someone will get there)

File: docs/apache-airflow/howto/usage-cli.rst [hunk 15]
```diff
@@ -217,6 +217,17 @@ You can use the ``--dry-run`` option to print the row counts in the primary tabl
 
 By default, ``db clean`` will archive purged rows in tables of the form ``_airflow_deleted__<table>__<timestamp>``.  If you don't want the data preserved in this way, you may supply argument ``--skip-archive``.
 
+Export the purged records from the archive tables
+-------------------------------------------------
+The ``db export-cleaned`` command exports the contents of the archived tables, created by the ``db clean`` command,
+to a specified format, by default to a CSV file. The exported file will contain the records that were purged from the
+primary tables during the ``db clean`` process.
+
+You can specify the export format using ``--export-format`` option. The default format is csv.
```
Comment (norm) [inline docs/apache-airflow/howto/usage-cli.rst]: Default … or only? If you fake add `json` as a second option, is `csv` still chosen by default?

File: docs/apache-airflow/howto/usage-cli.rst [hunk 16]
```diff
@@ -217,6 +217,17 @@ You can use the ``--dry-run`` option to print the row counts in the primary tabl
...
+
+You must also specify the location of the path to which you want to export the data using ``--output-path`` option.
```
Comment (norm) [inline docs/apache-airflow/howto/usage-cli.rst]: Worth mentioning that the dir must exist here?

**Review by norm — CHANGES_REQUESTED** [review]

File: docs/apache-airflow/howto/usage-cli.rst (hunk 15 above)
Comment (ephraimbuddy) [reply docs/apache-airflow/howto/usage-cli.rst]: We want to add more options in the future, so I chose default to be future proof

File: airflow/utils/db_cleanup.py (hunk 14 above)
Comment (ephraimbuddy) [reply airflow/utils/db_cleanup.py:132]: Actually, this is for future works that would add extra format options, so for json, we could say `if export_format=="json"`. we have protection on the choices using argparse `choices`. Do you suggest I add the message?

File: docs/apache-airflow/howto/usage-cli.rst (hunk 15 above)
Comment (jedcunningham) [reply docs/apache-airflow/howto/usage-cli.rst]: Let me rephrase, as I have the same question (I think).

``export-format`` isn't required, but also has no default. So you must pass `csv`, and it's also the only option.

I think having `csv` be the actual default is okay, but we need to mark it as such (then we wouldn't need to mark it as required). We should also list out the valid options here, which is only csv today.

Comment (ephraimbuddy) [reply docs/apache-airflow/howto/usage-cli.rst]: Ah, I get what @norm  and what you also mean: There's no default in code but here. Thanks, I will update it and update here too

**Review by norm — APPROVED** [review]
